keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,424 @@
+|6  |[scheduler.go:715](https://github.com/kubernetes/kubernetes/blob/15c3f1b11/staging/src/k8s.io/client-go/tools/record/event.go#L278)      |0.33%     |99.03%  |
+|7  |garbagecollector.go:404     |0.32%     |99.35%  |
+|8  |garbagecollector.go:517     |0.32%     |99.67%  |
+|9  |node_authorizer.go:197      |0.03%     |99.70%  |
+|10 |controller_utils.go:603     |0.03%     |99.73%  |
+|11 |replica_set.go:561          |0.03%     |99.76%  |
+|12 |trace.go:116                |0.02%     |99.78%  |
+|13 |deployment_controller.go:484|0.02%     |99.80%  |
+|14 |deployment_controller.go:575|0.02%     |99.82%  |
+|15 |endpoints_controller.go:340 |0.02%     |99.84%  |
+|16 |cloud_cidr_allocator.go:233 |0.02%     |99.86%  |
+|17 |ttl_controller.go:271       |0.01%     |99.87%  |
+|18 |replica_set.go:597          |0.01%     |99.88%  |
+|19 |replica_set.go:225          |0.01%     |99.88%  |
+|20 |scheduler.go:604            |0.01%     |99.89%  |
+|21 |factory.go:431              |0.01%     |99.90%  |
+|22 |kubelet_getters.go:173      |0.01%     |99.90%  |
+
+
+### Example of migrating klog call
+
+Before migration:
+```go
+klog.Infof("%s %s: (%v) %v%v%v [%s %s]", rl.req.Method, rl.req.RequestURI, latency, rl.status, rl.statusStack, rl.addedInfo, rl.req.UserAgent(), rl.req.RemoteAddr)
+```
+
+After migration:
+```go
+klog.InfoS("Access", "method", rl.req.Method, "uri", rl.req.RequestURI, "latency", latency, "status", rl.status, "agent", rl.req.UserAgent(), "addr", rl.req.RemoteAddr)
+```
+
+## Alternatives
+
+#### Just write guideline and update log messages
+
+* Doesn’t guarantee uniformity of log messages, is error prone and vulnerable to drifting
+* Doesn’t improve log ingestion into external systems
+* Doesn’t provide any value when switching to JSON output format
+
+#### Replace klog with some other structured logging library
+
+* Requires replacing all calls to klog with calls to a new logging library which makes this change hard to implement and merge into source repository.
+* Large percentage of the logs cannot be migrated to structured logging completely automatically, therefore full migration would require a significant amount of manual work and increase the risk of this change being abandoned.
+* Makes rollback of the change very hard to implement.
+
+#### Use glogr instead of proposed message structure
+
+* Glogr uses JSON serialization for both keys and values
+* Significantly increases log volume
+* Lack of support of warning and fatal severity levels
+
+## Code organisation / infrastructure needed
+
+As a result of this effort we are expecting creation of:
+* `k8s.io/component-base/metadata` - new module with utility functions for extracting metadata from k8s objects.
+
+## Production Readiness Review Process
+
+**Feature enablement and rollback**
+* How can this feature be enabled / disabled in a live cluster? **Changing logging format in control plane components would require recreation of cluster**
```

Comment:
Why recreation of a cluster? Why restarting a given component with a different option won't be enough?

Reply:
I guess it's a miswording here - it seems it just requires changing the flag (and restarting component?)

Reply:
Yes, when it comes to switching a log output format to json changing a flag value and restating a given component is sufficient.